kernel/kernel-api/src/main/java/io/delta/kernel/internal/icebergcompat/IcebergWriterCompatV3MetadataValidatorAndUpdater.java

```diff
@@ -0,0 +1,159 @@
+   * valid. Currently, the changes we support are
+   *
+   * <ul>
+   *   <li>No change in enablement (true to true or false to false)
+   *   <li>Enabling but only on a new table (false to true)
+   * </ul>
+   *
+   * The changes that we do not support and for which we throw an {@link KernelException} are
+   *
+   * <ul>
+   *   <li>Disabling on an existing table (true to false)
+   *   <li>Enabling on an existing table (false to true)
+   * </ul>
+   */
+  public static void validateIcebergWriterCompatV3Change(
+      Map<String, String> oldConfig, Map<String, String> newConfig, boolean isNewTable) {
+    validateIcebergWriterCompatChange(
+        oldConfig, newConfig, isNewTable, TableConfig.ICEBERG_WRITER_COMPAT_V3_ENABLED);
+  }
+
+  /**
+   * Validate and update the given Iceberg Writer Compat V3 metadata.
+   *
+   * @param newMetadata Metadata after the current updates
+   * @param newProtocol Protocol after the current updates
+   * @return The updated metadata if the metadata is valid and updated, otherwise empty.
+   * @throws UnsupportedOperationException if the metadata is not compatible with Iceberg Writer V3
+   *     requirements
+   */
+  public static Optional<Metadata> validateAndUpdateIcebergWriterCompatV3Metadata(
+      boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+    return validateAndUpdateIcebergWriterCompatMetadata(
+        isCreatingNewTable, newMetadata, newProtocol, INSTANCE);
+  }
+
+  /// //////////////////////////////////////////////////////////////////////////////
+  /// Define the compatibility and update checks for icebergWriterCompatV3       ///
+  /// //////////////////////////////////////////////////////////////////////////////
+
+  private static final IcebergWriterCompatV3MetadataValidatorAndUpdater INSTANCE =
+      new IcebergWriterCompatV3MetadataValidatorAndUpdater();
+
+  private static final IcebergCompatRequiredTablePropertyEnforcer ICEBERG_COMPAT_V3_ENABLED =
+      new IcebergCompatRequiredTablePropertyEnforcer<>(
+          TableConfig.ICEBERG_COMPAT_V3_ENABLED,
+          (value) -> value,
+          "true",
+          (inputContext) ->
+              IcebergCompatV3MetadataValidatorAndUpdater.validateAndUpdateIcebergCompatV3Metadata(
+                  inputContext.isCreatingNewTable,
+                  inputContext.newMetadata,
+                  inputContext.newProtocol));
+
+  /**
+   * Current set of allowed table features. This may evolve as the protocol evolves. This includes
+   * the incompatible legacy features (invariants, changeDataFeed, checkConstraints,
+   * identityColumns, generatedColumns) because they may be present in the table protocol even when
+   * they are not in use. In later checks we validate that these incompatible features are inactive
+   * in the table. See the protocol spec for more details.
+   */
+  private static Set<TableFeature> ALLOWED_TABLE_FEATURES =
```

Comment:
plz combine the list with the refactoring and only put new v3 allowed table feature here.